jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java

```diff
@@ -392,83 +358,161 @@ public abstract File getFile()
-     * Get a resource from within this resource.
-     */
-    @Override
-    public Resource getResource(String path) throws IOException
+    @Deprecated(forRemoval = true)
+    public String[] list()
     {
-        return addPath(path);
-    }
+        try (DirectoryStream<Path> dir = Files.newDirectoryStream(getPath()))
+        {
+            List<String> entries = new ArrayList<>();
+            for (Path entry : dir)
+            {
+                String name = entry.getFileName().toString();
 
-    // FIXME: this appears to not be used
-    @SuppressWarnings("javadoc")
-    public Object getAssociate()
-    {
-        return _associate;
+                if (Files.isDirectory(entry))
+                {
+                    name += "/";
+                }
+
+                entries.add(name);
+            }
+            int size = entries.size();
+            return entries.toArray(new String[size]);
+        }
+        catch (DirectoryIteratorException e)
+        {
+            LOG.debug("Directory list failure", e);
+        }
+        catch (IOException e)
+        {
+            LOG.debug("Directory list access failure", e);
+        }
+        return null;
     }
 
-    // FIXME: this appear to not be used
-    @SuppressWarnings("javadoc")
-    public void setAssociate(Object o)
+    public Resource resolve(String subPath) throws IOException
     {
-        _associate = o;
+        // Check that the path is within the root,
+        // but use the original path to create the
+        // resource, to preserve aliasing.
+        if (URIUtil.canonicalPath(subPath) == null)
+            throw new MalformedURLException(subPath);
+
+        if (URIUtil.SLASH.equals(subPath))
+            return this;
+
+        // Sub-paths are always resolved under the given URI,
+        // we compensate for input sub-paths like "/subdir"
+        // where default resolve behavior would be to treat
+        // that like an absolute path.
+        while (subPath.startsWith(URIUtil.SLASH))
+            subPath = subPath.substring(1);
+
+        URI uri = getURI();
+        URI resolvedUri;
+        if (uri.isOpaque())
+        {
+            // The 'jar:file:/some/path/my.jar!/foo/bar' URI is opaque b/c the jar: scheme is not followed by //
+            // so we take the scheme-specific part (i.e.: file:/some/path/my.jar!/foo/bar) and interpret it as a URI,
+            // use it to resolve the subPath then re-prepend the jar: scheme before re-creating the URI.
+            String scheme = uri.getScheme();
+            URI subUri = URI.create(uri.getSchemeSpecificPart());
+            if (subUri.isOpaque())
+                throw new IllegalArgumentException("Unsupported doubly opaque URI: " + uri);
+
+            if (!subUri.getPath().endsWith(URIUtil.SLASH))
+                subUri = URI.create(subUri + URIUtil.SLASH);
+
+            URI subUriResolved = subUri.resolve(subPath);
+            resolvedUri = URI.create(scheme + ":" + subUriResolved);
```

Comment:
It's good to have this general implemantion, but it does not look efficient. Hence I think it may well be worthwhile to java JarFileResource that can optimize this operation.